kbncontent.go

```diff
@@ -0,0 +1,319 @@
+	doc = objx.New(doc)
+	err := deserializeSubPaths(doc)
+
+	if err != nil {
+		return VisualizationDescriptor{}, fmt.Errorf("failed to deserialize embedded JSON objects: %w", err)
+	}
+
+	soType, ok := doc["type"].(string)
+	if !ok {
+		return VisualizationDescriptor{}, errors.New("`type` in visualization is not present or is not a string")
+	}
+
+	desc := VisualizationDescriptor{
+		Doc:    doc,
+		SoType: soType,
+		Link:   "by_reference",
+	}
+
+	return desc, nil
+}
+
+// Given a dashboard state (unmarshalled JSON), report information about the by-value panels
+func DescribeByValueDashboardPanels(dashboard interface{}) (visDescriptions []VisualizationDescriptor, err error) {
+	var panelsValue *objx.Value
+	if dashboardMap, ok := dashboard.(map[string]interface{}); ok {
+		panelsValue = objx.Map(dashboardMap).Get("attributes.panelsJSON")
+	} else {
+		return nil, fmt.Errorf("dashboard of unexpected type %T", dashboard)
+	}
+
+	var panels []objx.Map
+	if panelsValue.IsStr() {
+		result, err := objx.FromJSONSlice(panelsValue.Str())
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse panels JSON: %w", err)
+		}
+
+		panels = result
+	} else if panelsValue.IsObjxMapSlice() {
+		panels = panelsValue.ObjxMapSlice()
+	} else {
+		return nil, fmt.Errorf("panelsJSON is of unexpected type %T. Expected string or map[string]interface{}", err)
+	}
+
+	for _, panel := range panels {
+		panelTypeValue := panel.Get("type")
+		if !panelTypeValue.IsStr() {
+			// no type, so by-reference
+			continue
+		}
+
+		panelType := panelTypeValue.Str()
+
+		// TODO - I ported these checks from JS, but I do not understand why they are necessary
+		// also, note that this logic needs to change to support saved searches whenever they become by-value
+		filterOut := true
+
+		if panel.Has("embeddableConfig.savedVis") && panelType == "visualization" {
+			filterOut = false
+		}
+
+		if panel.Has("embeddableConfig.attributes") && (panelType == "lens" || panelType == "map") {
+			filterOut = false
+		}
+
+		if !filterOut {
+			desc := VisualizationDescriptor{
+				Doc:    panel,
+				SoType: panelType,
+				Link:   "by_value",
+			}
+			visDescriptions = append(visDescriptions, desc)
+		}
+	}
+	return visDescriptions, nil
+}
+
+func GetDashboardTitle(dashboard interface{}) (string, error) {
+	m := objx.Map(dashboard.(map[string]interface{}))
+	return m.Get("attributes.title").Str(), nil
+}
+
+// A dashboard reference
+type Reference struct {
+	ID, Type, Name string
+}
+
+func toReferenceSlice(val interface{}) ([]Reference, error) {
+	vals, ok := val.([]interface{})
+	if !ok {
+		return nil, errors.New("conversion error to array")
+	}
+	var refs []Reference
+	for _, v := range vals {
+		r, ok := v.(map[string]interface{})
+		if !ok {
+			return nil, errors.New("conversion error to reference element")
+		}
+		ref := Reference{
+			ID:   r["id"].(string),
+			Type: r["type"].(string),
+			Name: r["name"].(string),
+		}
```

Comment:
These unchecked conversions to string can panic. Maybe we can benefit here from something like https://github.com/mitchellh/mapstructure
```suggestion
		var ref Reference
		err := mapstructure.Decode(v, &ref)
		if err != nil {
			return nil, errors.Errorf("conversion errror to reference element: %w", err)
		}
```

`Reference` type might need some additional tags (not sure, it seems to default to the lowercased names, and in this case they would be equivalent).
```
type Reference struct {
	ID string   `mapstructure:"id"
	Type string `mapstructure:"type"`
	Name string `mapstructure:"name"`
}
```